### PR TITLE
[lte][agw] Adding ue_reported_usage session metric

### PR DESCRIPTION
**Pull request by ardzoht**
## Summary

- Adding both UL and DL session data usage metrics on `SessionState`
- Refactored helper metrics methods for cpp modules in `orc8r/gateway/c/MetricsHelpers`

## Test Plan

- running make integ_test
- validating metrics are updated with eNB and UE setup:

```
name: "ue_reported_tx"
help: ""
type: COUNTER
metric {
  label {
    name: "IMSI"
    value: "IMSI001010000000082"
  }
  label {
    name: "apn"
    value: "oai.ipv4"
  }
  label {
    name: "msisdn"
    value: ""
  }
  counter {
    value: 2098751.0
  }
}

name: "ue_reported_rx"
help: ""
type: COUNTER
metric {
  label {
    name: "IMSI"
    value: "IMSI001010000000082"
  }
  label {
    name: "apn"
    value: "oai.ipv4"
  }
  label {
    name: "msisdn"
    value: ""
  }
  counter {
    value: 141933908.0
  }
}
```

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 1]
```diff
@@ -1530,4 +1542,18 @@ void SessionState::update_bearer_deletion_req(
...
+  auto sid    = get_config().common_context.sid().id();
+  auto msisdn = get_config().common_context.msisdn();
+  auto apn    = get_config().common_context.apn();
+  increment_counter(
+      "ue_reported_tx", bytes_tx, size_t(3), LABEL_IMSI, sid.c_str(), LABEL_APN,
+      apn.c_str(), LABEL_MSISDN, msisdn.c_str());
+  increment_counter(
+      "ue_reported_rx", bytes_rx, size_t(3), LABEL_IMSI, sid.c_str(), LABEL_APN,
+      apn.c_str(), LABEL_MSISDN, msisdn.c_str());
+}
```
Comment (xjtian) [inline lte/gateway/c/session_manager/SessionState.cpp:1724]: What's the difference between these and the `ue_traffic` counter which already has labels for direction? Per-UE timeseries are extremely expensive to store and query on Prometheus at any sort of scale and we should avoid adding more unless absolutely necessary

Comment (ardzoht) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: `ue_traffic` use case is more for tracking and monitoring through a specific rating group and policy rule, these two are meant to cover session data usage for both LTE and CWF use cases (CWF currently gets UL/DL throughput metrics through the AP which seems to be a weak method). After we uniform these and confirm it works for both for our use cases then we can look to remove the ones in radius / CWF side.

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 2]
```diff
@@ -1530,4 +1542,18 @@ void SessionState::update_bearer_deletion_req(
   update.delete_req.mutable_eps_bearer_ids()->Add(
       bearer_id_by_policy_[PolicyID(policy_type, rule_id)]);
 }
+
+void SessionState::update_data_usage_metrics(
+    uint64_t bytes_tx, uint64_t bytes_rx) {
+  auto sid    = get_config().common_context.sid().id();
```
Comment (themarwhal) [inline lte/gateway/c/session_manager/SessionState.cpp]: nit: `const auto&` ?

**Review by themarwhal — APPROVED** [review]

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (xjtian) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: Can we feature-gate this to CWF only then? Will we be maintaining 2 sets of UE data usage timeseries in FWA if we don't?

I want to stress that it's *extremely* expensive to store and query these timeseries in Prometheus, particularly if label cardinality is high.

Comment (ardzoht) [reply lte/gateway/c/session_manager/SessionState.cpp:1686]: Ideally we should be using the same session metrics for FWA and CWF. But yeah I understand the concern on scalability and performance, I think we can start with these for CWF but in long term we might want to use these for both.

Comment (ardzoht) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: Ideally we should be using the same session metrics for FWA and CWF. But yeah I understand the concern on scalability and performance, I think we can start with these for CWF but in long term we might want to use these for both. Tagging @amarpad for his input

Comment (amarpad) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: Session counters per APN is also a specific request from some partners so we need something for FWA as well.

My higher level question here is should we have this be an event associated with session termination vs a counter.

Comment (xjtian) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: Is that a separate conversation, or the same thing? You mean report data usage only as part of a session termination event? It's slightly different semantics right? One is updated periodically (metric) and the other only set once per termination event. Also one will be queryable in grafana and the other not.

@ardzoht Let's use a single metric name for this and make direction tx/rx a label.

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 2 above)
Comment (ardzoht) [reply lte/gateway/c/session_manager/SessionState.cpp]: added `const` as getting the reference can lead to memory seg fault caused by a dangling reference

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (ardzoht) [reply lte/gateway/c/session_manager/SessionState.cpp:1724]: @xjtian Added a single metric and included the direction as label name

**Review by xjtian — APPROVED** [review]

**Pull request merged**